### PR TITLE
Create option for control objects to be issued repeatedly if a keyboard key is held.

### LINKED ISSUE #11569: Keyboard repeat option for certain ControlObjects
### Feature Description

It would be helpful if some control objects would be activated repeatedly if a keyboard control is held down.   These include:

* beatjump, for seeking quickly through a track
* beats_adjust_* and beats_translate_*  (this is where I want it the most, adjusting grids is very slow going tap tap tap tap tap tap right now)
* loop_double/halve
* rate_perm_ controls

To prevent accidental repetition we would implement this the standard way, with a long delay before first repetition, and then a shorter delay between each subsequent repetition.  

It would be a pretty small change to add an option to controlobject like setKeyboardRepeatable(bool) and then access that value in keyboardeventfilter and set up a small hashtable of timers (in case multiple keys are held) to manage the repetition of control object signal emissions.

### PR DESCRIPTION
**Pull request by ywwg**
Fixes https://github.com/mixxxdj/mixxx/issues/11569

Applies to:
* beats_adjust_faster, beats_adjust_slower
* beats_translate_earlier, beats_translate_later
* beat_next, beat_prev
* beatjump_size_halve, beatjump_size_double
* beatjump_forward, beatjump_backward
* loop_halve, loop_double
* beatjump_X_foward, beatjump_X_backward
* rate_perm_down, rate_perm_down_small, rate_perm_up, rate_perm_up_small

### CODE REVIEW

File: src/controllers/keyboard/keyboardeventfilter.h [hunk 1]
```diff
@@ -39,6 +40,18 @@ class KeyboardEventFilter : public QObject {
...
+
+    // Run through list of active keys to see if the pressed key is already active
+    // and is not a control that repeats when held.
+    bool shouldSkipHeldKey(int keyId) {
+        foreach (const KeyDownInformation& keyDownInfo, m_qActiveKeyList) {
+            if (keyDownInfo.keyId == keyId && !keyDownInfo.pControl->getKbdRepeatable()) {
+                return true;
+            }
+        }
+        return false;
```
Comment (Swiftb0y) [inline src/controllers/keyboard/keyboardeventfilter.h:52]:
```suggestion
        return std::any_of(
            m_qActiveKeyList.cbegin(), 
            m_qActiveKeyList.cend(), 
            [&](const KeyDownInformation& keyDownInfo){
                return keyDownInfo.keyId == keyId && !keyDownInfo.pControl->getKbdRepeatable();
        });
```

**Comment by ronso0** [comment]
Shouldn't this apply to all potmeter _up, _down, _up_small & _down_small controls?

**Comment by ywwg** [comment]
yup I can do that

**Comment by daschuer** [comment]
The code looks good for me and I can confirm the issue. 

I just wonder if this is a good approach, because until now, the keyboard keys are treated very similar as the controller keys and as GUI buttons. A press event becomes a NoteOn and the release events a NoteOff. The keyboard repetition has been intentionally disabled to keep that model working. 

After merge, a keyboard button has a special exclusive behavior. I think form this point of view a repetition logic inside the CO, that works independent from the source of the control event is the preferred solution. The user does probably not understand that we rely on the OS's repetition settings, that is only available for keyboards 

A case where these settings are miss-configured for Mixxx might be another setting.

Do you have interest to make this feature also work for controller and GUI buttons?

**Comment by ywwg** [comment]
> Do you have interest to make this feature also work for controller and GUI buttons?

I see what you mean -- if perhaps someone has linked the beatjump parameter to a button on a controller, they might want this behavior.  But I would prefer to write features in response to concrete need rather than hypothetical situations or abstract logical notions of consistency. In my experience I have never wanted this repeat behavior on a controller nor have I seen a feature request for it -- beatjump is usually bound to a clicking knob that is easy to spin.  I would want to see a concrete example where this might be a useful feature before coding it.

Whereas, when I am gridding and preparing my library with just my keyboard, or doing keyboard-only mixing, I definitely have wanted this keyboard repeat behavior, so I wanted to write this.

File: src/control/control.h [hunk 2]
```diff
@@ -198,6 +206,9 @@ class ControlDoublePrivate : public QObject {
     // User-visible, i18n description for what the control does.
     QString m_description;
 
+    // If true, this control will be issued repeatedly if the keyboard key is held.
+    bool m_kbd_repeatable;
```
[the remark this answers is not in the document]
Comment (ywwg) [reply src/control/control.h]: fixed

**Review by daschuer — APPROVED** [review]
LGTM, Thank you.

**Comment by ronso0** [comment]
Umm, why did this go to main?
It's nice usability improvement. Will check if this can be backported without hazzle.

**Pull request merged**